cunumeric/fft/fft.py

```diff
@@ -0,0 +1,879 @@
+        that axis is performed multiple times.  A one-element sequence means
+        that a one-dimensional FFT is performed.
+    norm : {"backward", "ortho", "forward"}, optional
+        Normalization mode (see `numpy.fft`). Default is "backward".
+        Indicates which direction of the forward/backward pair of transforms
+        is scaled and with what normalization factor.
+
+    Returns
+    -------
+    out : complex ndarray
+        The truncated or zero-padded input, transformed along the axes
+        indicated by `axes`, or the last two axes if `axes` is not given.
+
+    Notes
+    ------
+
+    See Also
+    --------
+    numpy.fft.fft2
+
+    Availability
+    --------
+    GPU
+    """
+    return fftn(a=a, s=s, axes=axes, norm=norm)
+
+
+@add_boilerplate("a")
+def fftn(a, s=None, axes=None, norm=None):
+    """
+    Compute the N-dimensional discrete Fourier Transform.
+
+    This function computes the *N*-dimensional discrete Fourier Transform over
+    any number of axes in an *M*-dimensional array by means of the Fast Fourier
+    Transform (FFT).
+
+    Parameters
+    ----------
+    a : array_like
+        Input array, can be complex.
+    s : sequence of ints, optional
+        Shape (length of each transformed axis) of the output
+        (``s[0]`` refers to axis 0, ``s[1]`` to axis 1, etc.).
+        This corresponds to ``n`` for ``fft(x, n)``.
+        Along any axis, if the given shape is smaller than that of the input,
+        the input is cropped.  If it is larger, the input is padded with zeros.
+        if `s` is not given, the shape of the input along the axes specified
+        by `axes` is used.
+    axes : sequence of ints, optional
+        Axes over which to compute the FFT.  If not given, the last ``len(s)``
+        axes are used, or all axes if `s` is also not specified.
+        Repeated indices in `axes` means that the transform over that axis is
+        performed multiple times.
+    norm : {"backward", "ortho", "forward"}, optional
+        Normalization mode (see `numpy.fft`). Default is "backward".
+        Indicates which direction of the forward/backward pair of transforms
+        is scaled and with what normalization factor.
+
+    Returns
+    -------
+    out : complex ndarray
+        The truncated or zero-padded input, transformed along the axes
+        indicated by `axes`, or by a combination of `s` and `a`,
+        as explained in the parameters section above.
+
+    Notes
+    ------
+
+    See Also
+    --------
+    numpy.fft.fftn
+
+    Availability
+    --------
+    GPU
+    """
+    if a.dtype == np.float32:
+        a = a.astype(np.complex64)
+    elif a.dtype == np.float64:
+        a = a.astype(np.complex128)
+
+    fft_type = None
+    if a.dtype == np.complex128:
+        fft_type = FFTCode.FFT_Z2Z
+    elif a.dtype == np.complex64:
+        fft_type = FFTCode.FFT_C2C
+    else:
+        raise TypeError("FFT input not supported (missing a conversion?)")
```

Comment:
What about "Input type not supported. fftn supports np.float32, np.float64, np.complex64 and np.complex128?" ?